internal/bgp/bgpfrr/frr.go

```diff
@@ -0,0 +1,242 @@
+package bgpfrr
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"go.universe.tf/metallb/internal/bgp"
+)
+
+// As the MetalLB controller should handle messages synchronously, there should
+// no need to lock this data structure. TODO: confirm this.
+type frrState struct {
+	sessions map[string]*Session
+}
+
+var state *frrState
+
+func init() {
+	state = &frrState{
+		sessions: map[string]*Session{},
+	}
+}
+
+type Session struct {
+	myASN      uint32
+	routerID   net.IP // May be nil, meaning "derive from context"
+	myNode     string
+	addr       string
+	srcAddr    net.IP
+	asn        uint32
+	holdTime   time.Duration
+	logger     log.Logger
+	password   string
+	advertised map[string]*bgp.Advertisement
+}
+
+// Create a variable for os.Hostname() in order to make it easy to mock out
+// in unit tests.
+var osHostname = os.Hostname
+
+// sessionName() defines the format of the key of the 'sessions' map in
+// the 'frrState' struct.
+func sessionName(myAddr string, myAsn uint32, addr string, asn uint32) string {
+	return fmt.Sprintf("%d@%s-%d@%s", asn, addr, myAsn, myAddr)
+}
+
+func validate(adv *bgp.Advertisement) error {
+	if adv.Prefix.IP.To4() != nil {
+		if adv.NextHop != nil && adv.NextHop.To4() == nil {
+			return fmt.Errorf("next-hop must be IPv4, got %q", adv.NextHop)
+		}
+	} else if adv.Prefix.IP.To16() != nil {
+		if adv.NextHop != nil && adv.NextHop.To16() == nil {
+			return fmt.Errorf("next-hop must be IPv6, got %q", adv.NextHop)
+		}
+	} else {
+		return fmt.Errorf("unable to validate IP address")
+	}
+
+	if len(adv.Communities) > 63 {
+		return fmt.Errorf("max supported communities is 63, got %d", len(adv.Communities))
+	}
+	return nil
+}
+
+func (s *Session) Commit() error {
```

Comment:
I know I proposed the `Commit` name, but maybe `Flush` is a better name for it?
Commit reminds me of a db transaction, but here we are not really starting a transaction>